storage/repair.go

```diff
@@ -0,0 +1,283 @@
+		origin   = session.Origin()
+		replicas = session.Replicas()
+	)
+
+	metadata := repair.NewReplicaMetadataComparer(replicas)
+
+	// Add local metadata
+	localMetadata, _ := shard.FetchBlocksMetadata(ctx, math.MaxInt64, 0, true, true)
+	localIter := block.NewFilteredBlocksMetadataIter(start, end, r.blockSize, localMetadata)
+	metadata.AddLocalMetadata(origin, localIter)
+
+	// Add peer metadata
+	peerIter, err := session.FetchBlocksMetadataFromPeers(namespace, shard.ID(), start, end, r.blockSize)
+	if err != nil {
+		return err
+	}
+	if err := metadata.AddPeerMetadata(peerIter); err != nil {
+		return err
+	}
+
+	// NB(xichen): only record the differences for now
+	r.recordFn(namespace, shard, metadata.Compare())
+
+	return nil
+}
+
+// TODO(xichen): log the actual differences once we have an idea of the magnitude of discrepancies
+func (r shardRepairer) recordDifferences(
+	namespace string,
+	shard databaseShard,
+	diffRes repair.MetadataComparisonResult,
+) {
+	var (
+		shardScope = r.scope.Tagged(map[string]string{
+			"namespace": namespace,
+			"shard":     strconv.Itoa(int(shard.ID())),
+		})
+		totalScope        = shardScope.Tagged(map[string]string{"resultType": "total"})
+		sizeDiffScope     = shardScope.Tagged(map[string]string{"resultType": "sizeDiff"})
+		checksumDiffScope = shardScope.Tagged(map[string]string{"resultType": "checksumDiff"})
+	)
+
+	// Record total number of series and total number of blocks
+	totalScope.Counter("series").Inc(diffRes.NumSeries)
+	totalScope.Counter("blocks").Inc(diffRes.NumBlocks)
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", diffRes.NumSeries),
+		xlog.NewLogField("numBlocks", diffRes.NumBlocks),
+	).Infof("repair total count")
+
+	// Record size differences
+	sizeDiffSeries := diffRes.SizeDifferences.Series()
+	numBlocks := 0
+	for _, series := range sizeDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	sizeDiffScope.Counter("series").Inc(int64(len(sizeDiffSeries)))
+	sizeDiffScope.Counter("blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", len(sizeDiffSeries)),
+		xlog.NewLogField("numBlocks", numBlocks),
+	).Infof("repair size difference count")
+
+	// Record checksum differences
+	checksumDiffSeries := diffRes.ChecksumDifferences.Series()
+	numBlocks = 0
+	for _, series := range checksumDiffSeries {
+		numBlocks += len(series.Blocks())
+	}
+	checksumDiffScope.Counter("series").Inc(int64(len(checksumDiffSeries)))
+	checksumDiffScope.Counter("blocks").Inc(int64(numBlocks))
+	r.logger.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("shard", shard.ID()),
+		xlog.NewLogField("numSeries", len(checksumDiffSeries)),
+		xlog.NewLogField("numBlocks", numBlocks),
+	).Infof("repair checksum difference count")
```

Comment:
I don't think its too bad considering this only runs every two hours and each node owns 64 shards, or 192 lines / 2 hours. Don't feel too strongly either though.

Might also need to include hostname in the tags though, otherwise the metrics are going to be aggregated across hosts and not useful.